### DISCUSSION TITLE
Custom tools: How to capture the output from bash and change the selection?

### DISCUSSION THREAD

**Discussion by AntonRich** [zim-desktop-wiki/zim-desktop-wiki#1969] [Q&A] [answered]
I'm try to accomplish the following:
1. I have a text with a bunch of returns

Concatentation must traverse the list on the left to replace its nil
with the list on the right. So, thanks to concatenation, you will
traverse the accumulated list for every item in the original list.

2. I created a script in Elixir to transform a text with returns to a text without returns.
```elixir
case System.argv() do
  [text] ->
      text
      |> String.split(~r/\n/)
      |> Enum.join(" ")
      |> IO.write()
end
```
3. I execute the command in Command Tools

/home/anton/.asdf/shims/elixir ~/scripts/elixir/echo.exs %t

4. The result I'm getting is nothing (it might an empty string).

5. I've tried to look at examples https://github.com/jaap-karssenberg/zim-wiki/wiki/Groovy-script-custom-tool
and saw that bash being used as an intermediary, so I wrote a script

```bash
#!/bin/bash
STR=$(elixir ~/scripts/elixir/echo.exs "$1")
echo $STR
```
However I get nothing as well.

As I understand the output is not captured? How can it be?
Although I tested the bash script with **echo "Whatever"** and it worked. "Whatever" string has been returned in place of my current selection in Zim.

**Comment by introt** [comment, marked as answer]
1. In case I can't help you out with Elixir, feel free to try [my bash script](https://github.com/introt/zim-customtools/blob/16c4290b173c0b2c4bca6debf53bf5ef9110f037/newlines-to-spaces#L3) for the same problem.
2. Have you tried running your Elixir script in the terminal to make sure it works & outputs to `stdout`?
3. Custom tools aren't ran in a shell, so the `~` in `~/scripts/elixir/echo.exs` won't get expanded, thus nothing is output.
4. Imo custom tools should report a non-zero exit code here - feel free to open a feature request.
5. You used an absolute path for `elixir` at 3 - I'd assume it's not in your `$PATH`? Try to run your bash script in the terminal first.

Hope this helps!

**Reply by AntonRich** [reply]
1. Hats off to you, sir. Your solution worked like a charm. There is one caveat, I used it with tilda ~/scripts/zim-wiki/remove_returns.sh in custom tools.
2. I have tried to test the Elixir script beforehand with an input file as an argument, parsed it with File.read!() and it worked. There are two ways to output to stdout: IO.puts and IO.write. I tried them both. They worked.
3. I tried to run Elixir script giving the full path. It didn't work.
4. I'm not sure I how to phrase that feature request in a proper way.
5. I used an absolute because I was surprised by the error message from Zim: "no such file or directory: elixir". I use just 'elixir' in my shell and it works fine irregardless of the directory I'm in. I use ASDF version manager, maybe that's somehow plays the role.

Thank you for your help. I'll mark your answer as a solution.

**Reply by introt** [reply]
> I'm not sure I how to phrase that feature request in a proper way.

No worries, I can open the ticket.

> Thank you for your help. I'll mark your answer as a solution.

Don't mention it :) I like solving problems & helping people[^blog], so it's a win-win for me. However, I'm not fully satisfied without getting your Elixir program to run, so I went back in...

> There is one caveat, I used it with tilda

Apparently the first `~` in a command (`argv[0]`) is expanded, but ones in the latter arguments are not[^source].
    https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/a532c35e4b54c8fa1823a483dc005e9f52865e7e/zim/applications.py#L199-L203

> I use ASDF version manager, maybe that's somehow plays the role.

I checked the ASDF docs, and not only does it play a role, it runs the production:

![image](https://user-images.githubusercontent.com/26069787/156780435-ba7f040d-39fd-4dd3-9aa6-002dfb008b34.png)

When you launch an interactive shell, your [configuration files](https://www.landoflinux.com/linux_bash_configuration_files.html) are read and executed. That configuration doesn't expand to Zim custom tools nor the non-interactive shell that executes bash scripts. You can verify this by running the following custom tool, set to replace the current selection:

    #!/usr/bin/env bash
    echo PATH: $PATH # doesn't include ~/.asdf/shims
    which elixir 2>&1 # "which: no elixir in $PATH" (redirected stderr to stdout)
    /home/anton/.asdf/shims/elixir /scripts/elixir/echo.exs "test" 2>&1 # No such file or directory OR some asdf error
    echo $? # exit code 127 (command not found)

My apologies for doubting your Elixir program. I installed Elixir using the system-wide package manager and the script worked as expected: `elixir /home/introt/echo.exs %t`. Prepending the line `#!/usr/bin/env elixir`, marking the file as an executable, and using `/home/introt/echo.exs %t` also worked.

So, if you want to run Elixir scripts from Zim, you could

1. Use a bash wrapper script which initializes the environment. Does involve quite a bit of overhead.
2. Install Elixir system wide. It should get overridden by asdf on interactive sessions.
3. Find out where asdf is storing the real executables, and hope they don't change.
4. Add a `.tool-versions` file and the necessary environment variables to Zim's env.

These exercises are left for the reader.

Happy Hacking!

[^blog]: Some unwarranted lore: `introt` was supposed to be the name of a blog.

[^source]: Source: original research, it was easier to try it than sift through all the code :)
